fix pulling full vectors from the primary key when column families are in use

mw5h opened the pull request:
#### vecindex: delete Manager.GetWithDesc() method

This method was added because backfilling indexes are not considered
deletable. In retrospect, there's no reason why we need to restrict an
internal interface this way and deleting it removes a fair bit of
complexity from upcoming changes to the vector store.

Informs: #146046
Release note: None

#### fetcher: return spanID from NextRowDecoded()

Previously, Fetcher.NextRowDecoded() discarded the spanID it got from
Fetcher.NextRow(). This information is needed for a future use of
NextRowDecoded(), so this patch passes it through. Existing users ignore
this return value.

Informs: #146046
Release note: None

#### vecstore: use span.Builder in implementation of GetFullVectors

Previously, the vector store stored the table and index descriptors for
the vector index it was associated with. Since the store itself did not
have a lease on this descriptor and would outlive the transaction that
caused the store to be created, we could end up using a stale table
descriptor for vector operations.

This patch moves table information into the vector transaction in the
form of an index fetch spec created by a leaseholder on the table. The
vecindex transaction is expected to have a lifetime shorter than the
lease held by the caller, so the descriptors that generate this fetch
spec will not go stale.

Addtionally, we generate the needed family IDs for a span.Splitter(), to
avoid reading families that we don't require. This is also included with
the fetch spec.

This patch also:
* Introduces a proto buffer to hold the index fetch spec and family IDs.
* Introduces a function to initialize this proto buffer.
* Introduces a testing API into the vecstore so that tests can provide
  fake family descriptors to vecstore.

Fixes: #146046
Release note (bug fix): A bug where using column families on tables
with vector indexes would cause the index to fail to return results has
been fixed.

#### vecstore: correctly handle composite pk columns

Previously, the vector store would blindly use primary key column bytes
encoded in a vector index key to rebuild the primary key when retrieving
full vector values. SpanBuilder will attempt to reused already encoded
datums when building a span, but it can't do so if the encoded datum was
encoded in a different direction than required for the span being built.
For non-composite columns this was not an issue because they have all
the data needed to re-encode in the key value itself. For composite
column, the data encoded in the key is a key analogue and can't be used
to reconstruct the original column value, which is stored as a suffix
column.

For vector search, the search processor has access to the ValueBytes
array as well as the key, so it's able to rebuild column values as
needed. However, the fixup processor only has access to the key bytes,
which requires a bit of trickery.

In this patch, we disallow assigning directionality to vector index key
columns. The mechanics of similarity search preclude scannign prefixes,
so this was always a nonsensical thing to do. We now disallow it
entirely and take the further step to mechanically switch the
directionality of vector index prefix columns to match the primary key,
ensuring that the encoded value of these shared columns is always
directly usable.

We also move the logic for decoding and rebuilding the primary key into
the vector store so that it can be shared between fixups and the vector
search processor. Note that there is an additional composite key decoder
available for the vector search processor because it does indeed need to
decode the primary key bytes back to column values.

Along the way, we:
*  Remove ASC/DESC from vector index columns.
*  Introduce a new index type method to describe indexes with
   non-directional prefixes.
*  Alter SHOW commands to no longer show vector index column
   directionality.
*  Introduce a bunch of test-gated code to ensure everything I said
   above is in fact true and remains so in the future.

Informs: #146046
Release note (sql change): It is no longer permitted to assign
directionality to any vector index column. Prefix columns (those columns
before the vector column in a compound vector index key) are not
scannable in a vector index, so directionality isn't relevant to them.

mw5h committed:
vecindex: delete Manager.GetWithDesc() method

This method was added because backfilling indexes are not considered
deletable. In retrospect, there's no reason why we need to restrict an
internal interface this way and deleting it removes a fair bit of
complexity from upcoming changes to the vector store.

Informs: #146046
Release note: None

mw5h committed:
vecstore: use span.Builder in implementation of GetFullVectors

Previously, the vector store stored the table and index descriptors for
the vector index it was associated with. Since the store itself did not
have a lease on this descriptor and would outlive the transaction that
caused the store to be created, we could end up using a stale table
descriptor for vector operations.

This patch moves table information into the vector transaction in the
form of an index fetch spec created by a leaseholder on the table. The
vecindex transaction is expected to have a lifetime shorter than the
lease held by the caller, so the descriptors that generate this fetch
spec will not go stale.

Addtionally, we generate the needed family IDs for a span.Splitter(), to
avoid reading families that we don't require. This is also included with
the fetch spec.

This patch also:
* Introduces a proto buffer to hold the index fetch spec and family IDs.
* Introduces a function to initialize this proto buffer.
* Introduces a testing API into the vecstore so that tests can provide
  fake family descriptors to vecstore.

Fixes: #146046
Release note (bug fix): A bug where using column families on tables
with vector indexes would cause the index to fail to return results has
been fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147307)
<!-- Reviewable:end -->

andy-kimball commented on pkg/sql/backfill/backfill.go:
```diff
@@ -477,6 +477,10 @@ type VectorIndexHelper struct {
 	indexPrefix []byte
+	// tableDesc is the table descriptor for the table being backfilled.
+	tableDesc catalog.TableDescriptor
+	// indexDesc is the index descriptor for the index being backfilled.
```
Where is `indexDesc` being initialized and used? I see where `tableDesc` is being used, but not `indexDesc`.

andy-kimball commented on pkg/sql/vecindex/mutation_searcher.go:
```diff
@@ -36,9 +37,9 @@ type MutationSearcher struct {
 //
 // NOTE: The index is expected to come from a call to Manager.Get, and therefore
 // using a vecstore.Store instance.
-func (s *MutationSearcher) Init(idx *cspann.Index, txn *kv.Txn) {
```
It'd be good to add a comment that `tableDesc` is expected to be already leased by the caller.

mw5h replied:
Done.
<!-- Reviewable comment -OTDlepy77gU6su6rWAv:b-896fix -->
<!-- Sent from Reviewable.io -->

andy-kimball commented on pkg/sql/vecindex/searcher.go:
```diff
@@ -36,9 +37,11 @@ type Searcher struct {
 //
 // NOTE: The index is expected to come from a call to Manager.Get, and therefore
 // using a vecstore.Store instance.
-func (s *Searcher) Init(idx *cspann.Index, txn *kv.Txn, baseBeamSize, maxResults int) {
```
Similar comment to above.

mw5h replied:
Done.
<!-- Reviewable comment -OTDlfw6CzHam68Bq03a:b-896fix -->
<!-- Sent from Reviewable.io -->

andy-kimball commented on pkg/sql/vecindex/vecstore/store.go:
```diff
@@ -110,7 +93,7 @@ func New(
+	defaultCodec keys.SQLCodec,
 	tableID catid.DescID,
 	indexID catid.IndexID,
 ) (ps *Store, err error) {
```
If we're leasing a descriptor within the scope of `DescsTxn`, is the lease still held once we exit the `DescsTxn`? Is it valid to use it further down, e.g. calling `tableDesc.ExternalRowData`?

mw5h replied:
Thanks for that. I made a TODO on another branch to come back and look at that and then didn't actually do it. I would not expect any guarantee on tableDesc outside of the DescsTxn.

andy-kimball commented on pkg/sql/vecindex/mutation_searcher.go:
```diff
@@ -30,17 +31,21 @@ type MutationSearcher struct {
 // NOTE: The index is expected to come from a call to Manager.Get, and therefore
 // using a vecstore.Store instance.
-func (s *MutationSearcher) Init(idx *cspann.Index, txn *kv.Txn, tableDesc catalog.TableDescriptor) {
+func (s *MutationSearcher) Init(
```
NIT: usually we put `evalCtx` as the first param (or 2nd after `context.Context`).

mw5h commented on pkg/sql/vecindex/vecstore/store_txn.go:
```diff
@@ -87,6 +91,15 @@ func (tx *Txn) Init(
 
+	// Find the ordinal of the vector column in the table.
+	vectorColID := tx.indexDesc.VectorColumnID()
+	for i, col := range tx.tableDesc.PublicColumns() {
```
I'm pretty sure it's correct, but I agree that testing is better. I don't see how we could do a pk lookup on a non-public column since we need that column to be there and be populated before we can GetFullVectors().

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, with a nit and with the understanding that we'll add good testing around using vector indexes on tables that are undergoing schema changes. Also, I don't know enough about `SpanBuilder` / `SpanSplitter` to do a good review that we're using those as expected, so maybe have someone with more knowledge in that area to take a look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-:-ORMpLd_8E_xvviRi-Nj:b9utop)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/rowexec/vector_search.go` line 66 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-ORMn7pQ9dRKWhMoMI5R:-ORMn7pQ9dRKWhMoMI5S:bm47w5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f1bc2ee240a8e58b6387a212f4ac3c7b6de0343/pkg/sql/rowexec/vector_search.go#L66)):*
> ```Go
> 	searchBeamSize := int(flowCtx.EvalCtx.SessionData().VectorSearchBeamSize)
> 	maxResults := int(v.targetCount)
> 	v.searcher.Init(flowCtx.EvalCtx, idx, flowCtx.Txn, tabledesc.NewBuilder(&spec.Table).BuildImmutableTable(), searchBeamSize, maxResults)
> ```

NIT: Line is longer than 80 chars, I'd set the table descriptor in a local variable in the line above, since it's a lot of chars to do that inline in the call.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> Reviewed 2 of 4 files at r1, 16 of 16 files at r9, 11 of 12 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147307)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 42 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-ORPj9QG4wOPV1e6HNj9:-ORPj9QHEXl059PB9g9Z:b-ocrp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cca5bfba772efe298a523a21bfcb88a67e049f/pkg/sql/vecindex/vecstore/store_txn.go#L42)):*
> ```Go
> 	codec     partitionCodec
> 	tableDesc catalog.TableDescriptor
> 	indexDesc catalog.Index
> ```

What's stopping us from having the `vecindex.Searcher` and `vecindex.MutationSearcher` pass in the initialized `fullVecFetchSpec` and `VectorColumnID`? Could we avoid passing around the descriptor for the SQL search operators if we did that?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
This is great work!

> Reviewed 1 of 12 files at r10, 12 of 12 files at r11, 2 of 2 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-:-ORRkLojAYI90Lrg4kQA:bkm56jo)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 402 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-ORReg1U90hwKi8I1pPE:-ORReg1U90hwKi8I1pPF:b-abf6aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2af7c42e320879d24bdaee4408eee92ce923705b/pkg/sql/vecindex/vecstore/store_txn.go#L402)):*
> ```Go
> 		}
> 
> 		_, num, err := rowenc.DecodeKeyVals(tx.tmpEncDatums, tx.pkDir, ref.Key.KeyBytes)
> ```

I think we're missing the case when part of the primary key is in the `TreeKey` prefix rather than the `KeyBytes` suffix. See also the similar logic in the vector search operator: https://github.com/cockroachdb/cockroach/blob/24e1ef3ea350772fa6b1d6f56c22c91d93df8161/pkg/sql/rowexec/vector_search.go#L173-L183

We should also make sure to have a test for that case.
___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 412 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-ORRhN9W0Fc6rYf1F9aV:-ORRhN9W0Fc6rYf1F9aW:bb7ik8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc256161502af884e6e37ad336b2d37b5caa8c64/pkg/sql/vecindex/vecstore/store_txn.go#L412)):*
> ```Go
> 	var neededColOrdinals intsets.Fast
> 	neededColOrdinals.Add(tx.vectorColumnOrdinal)
> 	splitter := span.MakeSplitter(tx.tableDesc, tx.tableDesc.GetPrimaryIndex(), neededColOrdinals)
> ```

Along the same vein as my other comment, we could avoid having to pass the descriptors down here for the SQL path by using `MakeSplitterWithFamilyIDs`. The idea would be that when we have access to the descriptor during planning, we make a splitter and then propagate the family IDs it calculates. We do the same thing for JoinReaders: https://github.com/cockroachdb/cockroach/blob/24e1ef3ea350772fa6b1d6f56c22c91d93df8161/pkg/sql/distsql_physical_planner.go#L3375

And of course, the fixup path would have the descriptor here, so can make a splitter easily.


<!-- Sent from Reviewable.io -->

mw5h committed:
fetcher: return spanID from NextRowDecoded()

Previously, Fetcher.NextRowDecoded() discarded the spanID it got from
Fetcher.NextRow(). This information is needed for a future use of
NextRowDecoded(), so this patch passes it through. Existing users ignore
this return value.

Informs: #146046
Release note: None

mw5h commented:
RFAL

DrewKimball commented on pkg/sql/vecindex/vecstore/store_txn.go:
```diff
@@ -309,10 +333,128 @@ func (tx *Txn) SearchPartitions(
+func (d *pkDecoder) Decode(
+	treeKey cspann.TreeKey, keyBytes []byte,
+) (row rowenc.EncDatumRow, err error) {
+	decodeFromKey := func(keyCols []fetchpb.IndexFetchSpec_KeyColumn, keyBytes []byte) error {
```
nit: Feels like we could pull out a common helper from here and the vector search operator - but it's fine if that's just a TODO for now.

DrewKimball commented on pkg/sql/vecindex/vecstore/store_txn.go:
```diff
@@ -309,10 +333,128 @@ func (tx *Txn) SearchPartitions(
+		}
+		return nil
+	}
+	// Decode the index prefix columns, if any,
```
nit: comma at the end.

Also, maybe mention that we only need to do this if PK columns are included in the prefix.

DrewKimball reviewed (commented):
This looks good, apart from the potential issue with PK columns with composite encoding, maybe only a problem if the PK direction is different.

Are you leaning toward finishing up this approach, or pulling apart `TreeKey` and `KeyBytes`?

mw5h committed:
vecstore: correctly handle composite pk columns

Previously, the vector store would blindly use primary key column bytes
encoded in a vector index key to rebuild the primary key when retrieving
full vector values. SpanBuilder will attempt to reused already encoded
datums when building a span, but it can't do so if the encoded datum was
encoded in a different direction than required for the span being built.
For non-composite columns this was not an issue because they have all
the data needed to re-encode in the key value itself. For composite
column, the data encoded in the key is a key analogue and can't be used
to reconstruct the original column value, which is stored as a suffix
column.

For vector search, the search processor has access to the ValueBytes
array as well as the key, so it's able to rebuild column values as
needed. However, the fixup processor only has access to the key bytes,
which requires a bit of trickery.

In this patch, we disallow assigning directionality to vector index key
columns. The mechanics of similarity search preclude scannign prefixes,
so this was always a nonsensical thing to do. We now disallow it
entirely and take the further step to mechanically switch the
directionality of vector index prefix columns to match the primary key,
ensuring that the encoded value of these shared columns is always
directly usable.

We also move the logic for decoding and rebuilding the primary key into
the vector store so that it can be shared between fixups and the vector
search processor. Note that there is an additional composite key decoder
available for the vector search processor because it does indeed need to
decode the primary key bytes back to column values.

Along the way, we:
*  Remove ASC/DESC from vector index columns.
*  Introduce a new index type method to describe indexes with
   non-directional prefixes.
*  Alter SHOW commands to no longer show vector index column
   directionality.
*  Introduce a bunch of test-gated code to ensure everything I said
   above is in fact true and remains so in the future.

Informs: #146046
Release note (sql change): It is no longer permitted to assign
directionality to any vector index column. Prefix columns (those columns
before the vector column in a compound vector index key) are not
scannable in a vector index, so directionality isn't relevant to them.

fqazi reviewed (changes requested):
> Reviewed 2 of 18 files at r13, 1 of 32 files at r19, 6 of 6 files at r20, 23 of 26 files at r21, 17 of 17 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147307)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @DrewKimball, and @mw5h)

___
*[`pkg/sql/logictest/testdata/logic_test/vector_index` line 1069 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-OTDJN1U2SddN1r5yGP3:-OTDJN1U2SddN1r5yGP4:b-4xt7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f479f90f44836d97a342d9be6ef4723c400f4679/pkg/sql/logictest/testdata/logic_test/vector_index#L1069)):*
> ```
> DROP TABLE composite_pk
> 
> subtest end
> ```

Can we get a test with ALTER PRIMARY KEY that confirms the recreated index is usable
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 683 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-OTDIXok5ZHAEvqrZ8w-:-OTDIXok5ZHAEvqrZ8w0:bf9rea7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f479f90f44836d97a342d9be6ef4723c400f4679/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L683)):*
> ```Go
> 				} else if idx.Type == idxtype.VECTOR {
> 					for i, col := range inColumns {
> 						if col.columnID == ics.columnID {
> ```

Would this condition ever be true? The previous if statement confirmed ics was not already there in idxColIDs? inColumns and idxColIDs are kept in sync above.


<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147307)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @DrewKimball, and @mw5h)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 683 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-OTDIXok5ZHAEvqrZ8w-:-OTDJt3n1AS2qmya0_Va:bin04ri) ([raw file](https://github.com/cockroachdb/cockroach/blob/f479f90f44836d97a342d9be6ef4723c400f4679/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L683)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Would this condition ever be true? The previous if statement confirmed ics was not already there in idxColIDs? inColumns and idxColIDs are kept in sync above.
</blockquote></details>

I think we just want to adjust the direction adding these columns in above?


<!-- Sent from Reviewable.io -->

DrewKimball commented on pkg/sql/alter_primary_key.go:
```diff
@@ -845,10 +845,13 @@ func setKeySuffixAndStoredColumnIDsFromPrimary(
+				return errors.AssertionFailedf(
+					"indexed vector column cannot be part of the primary key")
+			}
+			tabledesc.UpdateVectorIndexPrefixColDirections(toAdd, primary)
```
nit: Should we do this after the loop?

DrewKimball commented on pkg/sql/logictest/testdata/logic_test/vector_index:
```diff
@@ -1015,7 +1018,7 @@ CREATE TABLE composite_pk (
   cust_id INT NOT NULL,
   embedding VECTOR(3),
-  PRIMARY KEY (cust_name, cust_id)
+  PRIMARY KEY (cust_name DESC, cust_id)
```
We should test the output of `SHOW CREATE TABLE ...` to make sure that vector indexes are displayed with no direction, even when the shared PK-prefix columns are descending.

mw5h replied:
Done.
<!-- Reviewable comment -OTDt4GfBO0izvvl6PII:b-896fix -->
<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
LGTM (pending Faizan's review)

mw5h commented on pkg/sql/vecindex/vecstore/store.go:
```diff
@@ -45,63 +49,41 @@ type Store struct {
+// NewWithLeasedDesc creates a Store for an index on the provided table descriptor
+// using the provided index descriptor. This is used in unit tests where full
+// vector index creation capabilities aren't necessarily available.
+func NewWithLeasedDesc(
```
Done.
<!-- Reviewable comment -OTDlh1O14u_qn2ZhSxX:b-896fix -->
<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147307)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @DrewKimball, and @fqazi)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 683 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-OTDIXok5ZHAEvqrZ8w-:-OTDkv_H0T9AJcdaOT_I:b-jo7nsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f479f90f44836d97a342d9be6ef4723c400f4679/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L683)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think we just want to adjust the direction adding these columns in above?
</blockquote></details>

This is maybe not the clearest code. The previous if statement catches the case where the PK column isn't in the index key, so we go down this route only when we know the PK column is also in the index prefix.
___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 42 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-ORPj9QG4wOPV1e6HNj9:-OTDlidQ2_9B3PR69nP9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0cca5bfba772efe298a523a21bfcb88a67e049f/pkg/sql/vecindex/vecstore/store_txn.go#L42)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

What's stopping us from having the `vecindex.Searcher` and `vecindex.MutationSearcher` pass in the initialized `fullVecFetchSpec` and `VectorColumnID`? Could we avoid passing around the descriptor for the SQL search operators if we did that?
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 402 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-ORReg1U90hwKi8I1pPE:-OTDlWeREEIvkL2e8Pln:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2af7c42e320879d24bdaee4408eee92ce923705b/pkg/sql/vecindex/vecstore/store_txn.go#L402)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think we're missing the case when part of the primary key is in the `TreeKey` prefix rather than the `KeyBytes` suffix. See also the similar logic in the vector search operator: https://github.com/cockroachdb/cockroach/blob/24e1ef3ea350772fa6b1d6f56c22c91d93df8161/pkg/sql/rowexec/vector_search.go#L173-L183

We should also make sure to have a test for that case.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 412 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-ORRhN9W0Fc6rYf1F9aV:-OTDlUQA1J9-Kj41E2ye:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc256161502af884e6e37ad336b2d37b5caa8c64/pkg/sql/vecindex/vecstore/store_txn.go#L412)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Along the same vein as my other comment, we could avoid having to pass the descriptors down here for the SQL path by using `MakeSplitterWithFamilyIDs`. The idea would be that when we have access to the descriptor during planning, we make a splitter and then propagate the family IDs it calculates. We do the same thing for JoinReaders: https://github.com/cockroachdb/cockroach/blob/24e1ef3ea350772fa6b1d6f56c22c91d93df8161/pkg/sql/distsql_physical_planner.go#L3375

And of course, the fixup path would have the descriptor here, so can make a splitter easily.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/vector_index` line 1069 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-OTDJN1U2SddN1r5yGP3:-OTDt4aQDvrYdKCYHHbf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f479f90f44836d97a342d9be6ef4723c400f4679/pkg/sql/logictest/testdata/logic_test/vector_index#L1069)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Can we get a test with ALTER PRIMARY KEY that confirms the recreated index is usable
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
Nice work <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 46 of 46 files at r23, 6 of 6 files at r24, 26 of 26 files at r25, 10 of 17 files at r26, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-:-OTDvzQK2tHR8nH39FsO:b-fekxoh)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @DrewKimball)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 683 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/147307#-OTDIXok5ZHAEvqrZ8w-:-OTDugShAsbC5z_-vwfj:b-nhssdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f479f90f44836d97a342d9be6ef4723c400f4679/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L683)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

This is maybe not the clearest code. The previous if statement catches the case where the PK column isn't in the index key, so we go down this route only when we know the PK column is also in the index prefix.
</blockquote></details>

Done.

Got it that makes sense. Maybe clarify the comment a tiny bit to help readability.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
> Reviewed 7 of 17 files at r26.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147307)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball and @DrewKimball)



<!-- Sent from Reviewable.io -->
